maro/simulator/scenarios/supply_chain/units/seller.py

```diff
@@ -171,9 +182,140 @@ class DataFileDemandSampler(SellerDemandSampler):
             . "sale_column", column name that will be treated as sale number (demand).
             . "datetime_column", column name that contains datetime, NOTE: we will parse it that ignore the time zone.
     """
+    def __init__(self, configs: dict, world: World) -> None:
+        super(PreprocessedFileDemandSampler, self).__init__(configs, world)
+
+        self._file_path = configs["file_path"]
+        self._preprocessed_file_path = get_preprocessed_file_path(configs["file_path"])
+
+        self._start_date_index: Optional[int] = None
+        start_date_time = self._world.configs.settings["start_date_time"]
+        if start_date_time is not None:
+            self._start_date_index = get_date_index(parse(start_date_time))
+
+        self._sku_column_name = configs.get("sku_column", "SKU")
+        self._price_column_name = configs.get("price_column", "Price")
+        self._sale_column_name = configs.get("sale_column", "Sales")
+        self._datetime_column_name = configs.get("datetime_column", "DT")
+
+        if not os.path.exists(self._preprocessed_file_path):
+            print(f"Preprocessed file {self._preprocessed_file_path} does not exist. Start preprocessing now.")
+            preprocess_file(self._file_path)
+
+        self._cache = collections.defaultdict(dict)
+
+
+class PreprocessedFileDemandSimpleSampler(PreprocessedFileDemandSampler):
+    """Load & cache all data when initializing.
+    """
+    def __init__(self, configs: dict, world: World) -> None:
+        super(PreprocessedFileDemandSimpleSampler, self).__init__(configs, world)
+        self._load_all_data()
+
+    def sample_demand(self, product_id: int, tick: int) -> int:
+        if tick not in self._cache or product_id not in self._cache[tick]:
+            return 0
+        return self._cache[tick][product_id].sales
+
+    def _load_all_data(self) -> None:
+        with open(self._preprocessed_file_path, "rt") as fp:
+            reader = DictReader(fp)
+
+            for row in tqdm(reader, desc=f"Loading data from {fp.name}"):
+                sku_name = row[self._sku_column_name]
+                if sku_name not in self._world.sku_name2id_mapping:
+                    continue
+
+                sales = int(row[self._sale_column_name])
+                price = float(row[self._price_column_name])
+                date_index = int(row[DATE_INDEX_COLUMN_NAME])
+
+                if self._start_date_index is None:
+                    self._start_date_index = date_index
+
+                # So one day one tick.
+                target_tick = date_index - self._start_date_index
+                sku = self._world.get_sku_by_name(sku_name)
+
+                if sku is not None:
+                    self._cache[target_tick][sku.id] = SkuRow(price, sales)
+                else:
+                    warnings.warn(f"{sku_name} not configured in config file.")
+
+
+class PreprocessedFileDemandStreamSampler(PreprocessedFileDemandSampler):
```

Comment:
Since the data are cached all the time. Is there any significant gaps between the Simple one and the Stream one for a whole episode?